src/realm/object-store/sync/sync_session.cpp

```diff
@@ -930,7 +950,7 @@ SyncSession::State SyncSession::state() const
 
 SyncSession::ConnectionState SyncSession::connection_state() const
 {
-    util::CheckedUniqueLock lock(m_state_mutex);
+    util::CheckedUniqueLock lock(m_connection_state_mutex);
```

Comment:
Using m_state_mutex here lead to a deadlock when using the synchronous wait for completion functions.